keccak256/src/gates/mixing.rs

```diff
@@ -23,18 +24,55 @@ pub struct MixingConfig<F> {
     base_conv_config: StateBaseConversion<F>,
     state: [Column<Advice>; 25],
     flag: Column<Advice>,
+    q_flag: Selector,
+    q_out_copy: Selector,
+    out_mixing: [Column<Advice>; 25],
 }
 
 impl<F: FieldExt> MixingConfig<F> {
     pub fn configure(
         meta: &mut ConstraintSystem<F>,
         table: FromBase9TableConfig<F>,
+        round_ctant_b9: Column<Advice>,
+        round_ctant_b13: Column<Advice>,
+        round_constants_b9: Column<Instance>,
+        round_constants_b13: Column<Instance>,
     ) -> MixingConfig<F> {
         // Allocate space for the flag column from which we will copy to all of
         // the sub-configs.
         let flag = meta.advice_column();
         meta.enable_equality(flag.into());
 
+        // Generate a selector that will always be active to avoid the
+        // PoisonedConstraint err due to no selectors being used in a
+        // constraint.
+        let q_flag = meta.selector();
+
+        meta.create_gate("Ensure flag consistency", |meta| {
+            let q_flag = meta.query_selector(q_flag);
+
+            let negated_flag = meta.query_advice(flag, Rotation::next());
+            let flag = meta.query_advice(flag, Rotation::cur());
+            // We do a trick which consists on multiplying an internal selector
+            // which is always active by the actual `negated_flag`
+            // which will then enable or disable the gate.
+            //
+            // Force that `flag + negated_flag = 1`.
+            // This ensures that flag = !negated_flag.
+            let flag_consistency =
+                (flag.clone() + negated_flag.clone()) - Expression::Constant(F::one());
+
+            // Define bool constraint for flags.
+            // Based on: `(1-flag) * flag = 0` only if `flag` is boolean.
+            let bool_constraint = |flag: Expression<F>| -> Expression<F> {
+                (Expression::Constant(F::one()) - flag.clone()) * flag
+            };
+
+            // Add a constraint that sums up the results of the two branches
+            // constraining it to be equal to `out_state`.
+            [q_flag * (flag_consistency + bool_constraint(flag) + bool_constraint(negated_flag))]
```

Comment:
We should split 3 polys here. Otherwise, the prover can witness values not satisfying individual polys but satisfying the summation poly.
Proof: let `x=flag`, `y=negated_flag`
flag_consistency: `x + y - 1`
bool_constraint for x: `x(1-x)`
bool_constraint for y: `y(1-y)`
summation: `x + y - 1 + x(1-x) + y(1-y) = 0`
falsifying example: `x=2 and y=1`

```suggestion
            [q_flag.clone() * flag_consistency, q_flag.clone() * bool_constraint(flag), q_flag * bool_constraint(negated_flag))]
```

Reply:
That makes sense!!! Thanks for catching this!!!!